chloggen/main.go

```diff
@@ -0,0 +1,237 @@
+	path := filepath.Join(ctx.unreleasedDir, cleanFileName(filename))
+	var pathWithExt string
+	switch ext := filepath.Ext(path); ext {
+	case ".yaml":
+		pathWithExt = path
+	case ".yml":
+		pathWithExt = strings.TrimSuffix(path, ".yml") + ".yaml"
+	case "":
+		pathWithExt = path + ".yaml"
+	default:
+		return fmt.Errorf("non-yaml extension: %s", ext)
+	}
+
+	templateBytes, err := os.ReadFile(ctx.templateYAML)
+	if err != nil {
+		return err
+	}
+	err = os.WriteFile(pathWithExt, templateBytes, os.FileMode(0755))
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Changelog entry template copied to: %s\n", pathWithExt)
+	return nil
+}
+
+func validate(ctx chlogContext) error {
+	if _, err := os.Stat(ctx.unreleasedDir); err != nil {
+		return err
+	}
+
+	entries, err := readEntries(ctx)
+	if err != nil {
+		return err
+	}
+	for _, entry := range entries {
+		if err = entry.Validate(); err != nil {
+			return err
+		}
+	}
+	fmt.Printf("PASS: all files in ./%s/ are valid\n", ctx.unreleasedDir)
+	return nil
+}
+
+func update(ctx chlogContext, version string, dry bool) error {
+	entries, err := readEntries(ctx)
+	if err != nil {
+		return err
+	}
+
+	if len(entries) == 0 {
+		return fmt.Errorf("no entries to add to the changelog")
+	}
+
+	chlogUpdate, err := generateSummary(version, entries)
+	if err != nil {
+		return err
+	}
+
+	if dry {
+		fmt.Printf("Generated changelog updates:")
+		fmt.Println(chlogUpdate)
+		return nil
+	}
+
+	oldChlogBytes, err := os.ReadFile(ctx.changelogMD)
+	if err != nil {
+		return err
+	}
+	chlogParts := bytes.Split(oldChlogBytes, []byte(insertPoint))
+	if len(chlogParts) != 2 {
+		return fmt.Errorf("expected one instance of %s", insertPoint)
+	}
+
+	chlogHeader, chlogHistory := string(chlogParts[0]), string(chlogParts[1])
+
+	var chlogBuilder strings.Builder
+	chlogBuilder.WriteString(chlogHeader)
+	chlogBuilder.WriteString(insertPoint)
+	chlogBuilder.WriteString(chlogUpdate)
+	chlogBuilder.WriteString(chlogHistory)
+
+	tmpMD := ctx.changelogMD + ".tmp"
+	tmpChlogFile, err := os.Create(tmpMD)
+	if err != nil {
+		return err
+	}
+	if _, err := tmpChlogFile.WriteString(chlogBuilder.String()); err != nil {
+		return err
+	}
+	if err := tmpChlogFile.Close(); err != nil {
+		return err
+	}
+
+	if err := os.Rename(tmpMD, ctx.changelogMD); err != nil {
+		return err
+	}
+
+	fmt.Printf("Finished updating %s\n", ctx.changelogMD)
+
+	return deleteEntries(ctx)
+}
+
+func readEntries(ctx chlogContext) ([]*Entry, error) {
+	entryYAMLs, err := filepath.Glob(filepath.Join(ctx.unreleasedDir, "*.yaml"))
+	if err != nil {
+		return nil, err
+	}
+
+	entries := make([]*Entry, 0, len(entryYAMLs))
+	for _, entryYAML := range entryYAMLs {
+		if filepath.Base(entryYAML) == filepath.Base(ctx.templateYAML) {
+			continue
+		}
+
+		fileBytes, err := os.ReadFile(entryYAML)
+		if err != nil {
+			return nil, err
+		}
+
+		entry := &Entry{}
+		if err = yaml.Unmarshal(fileBytes, entry); err != nil {
+			return nil, err
+		}
```

Comment:
nit: do we want to fall at the first error, or collect them all, so we have a list of what went wrong?

Reply:
Collecting them all would be great. In practice it doesn't matter much as long as validation runs regularly. A PR will typically only add one entry, so there are a very small number of possible errors to work through.